controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1707,30 +1707,35 @@ func (r *BareMetalHostReconciler) updateEventHandler(e event.UpdateEvent) bool {
 }
 
 // SetupWithManager registers the reconciler to be run by the manager
-func (r *BareMetalHostReconciler) SetupWithManager(mgr ctrl.Manager, preprovImgEnable bool) error {
+func (r *BareMetalHostReconciler) SetupWithManager(mgr ctrl.Manager, preprovImgEnable bool, options controller.Options) error {
 
-	maxConcurrentReconciles := runtime.NumCPU()
-	if maxConcurrentReconciles > 8 {
-		maxConcurrentReconciles = 8
-	}
-	if maxConcurrentReconciles < 2 {
-		maxConcurrentReconciles = 2
-	}
-	if mcrEnv, ok := os.LookupEnv("BMO_CONCURRENCY"); ok {
-		mcr, err := strconv.Atoi(mcrEnv)
-		if err != nil {
-			return errors.Wrap(err, fmt.Sprintf("BMO_CONCURRENCY value: %s is invalid", mcrEnv))
+	maxConcurrentReconciles := 0
+	if options.MaxConcurrentReconciles > 0 {
+		maxConcurrentReconciles = options.MaxConcurrentReconciles
+	} else {
+		maxConcurrentReconciles = runtime.NumCPU()
+
+		if maxConcurrentReconciles > 8 {
+			maxConcurrentReconciles = 8
+		}
+		if maxConcurrentReconciles < 2 {
+			maxConcurrentReconciles = 2
 		}
-		if mcr > 0 {
-			ctrl.Log.Info(fmt.Sprintf("BMO_CONCURRENCY of %d is set via an environment variable", mcr))
-			maxConcurrentReconciles = mcr
+		if mcrEnv, ok := os.LookupEnv("BMO_CONCURRENCY"); ok {
+			mcr, err := strconv.Atoi(mcrEnv)
```

Comment:
Why not move all of this out into a common function, so that each controller can use the same logic?

Reply:
Fixed